query/order.go

```diff
@@ -1,66 +1,48 @@
 package query
 
 import (
-	"sort"
+	"bytes"
+	"strings"
 )
 
 // Order is an object used to order objects
 type Order interface {
-
-	// Sort sorts the Entry slice according to
-	// the Order criteria.
-	Sort([]Entry)
+	Compare(a, b Entry) int
 }
 
-// OrderByValue is used to signal to datastores they
-// should apply internal orderings. unfortunately, there
-// is no way to apply order comparisons to interface{} types
-// in Go, so if the datastore doesnt have a special way to
-// handle these comparisons, you must provide an Order
-// implementation that casts to the correct type.
-type OrderByValue struct {
-	TypedOrder Order
+// OrderByFunction orders the results based on the result of the given function.
+type OrderByFunction func(a, b Entry) int
+
+func (o OrderByFunction) Compare(a, b Entry) int {
+	return o(a, b)
 }
 
-func (o OrderByValue) Sort(res []Entry) {
-	if o.TypedOrder == nil {
-		panic("cannot order interface{} by value. see query docs.")
-	}
-	o.TypedOrder.Sort(res)
+// OrderByValue is used to signal to datastores they should apply internal
+// orderings.
+type OrderByValue struct{}
+
+func (o OrderByValue) Compare(a, b Entry) int {
+	return bytes.Compare(a.Value, b.Value)
 }
 
 // OrderByValueDescending is used to signal to datastores they
-// should apply internal orderings. unfortunately, there
-// is no way to apply order comparisons to interface{} types
-// in Go, so if the datastore doesnt have a special way to
-// handle these comparisons, you are SOL.
-type OrderByValueDescending struct {
-	TypedOrder Order
-}
+// should apply internal orderings.
+type OrderByValueDescending struct{}
```

Comment:
Not related to the changes introduced herein, but since were making breaking changes, I think descending could just be an option of `OrderByKey` and `OrderByValue`.